src/glossary.adoc

```diff
@@ -0,0 +1,116 @@
+
+definition::
+The specification of a new entity. Most (but not all) definitions are declarations.
+
+directive::
+A unit of data in an Ion document that affects the encoding context, and thus the way the
+document’s data is decoded. In Ion 1.0 there are two directives: the Ion Version Marker, and
+top-level `$ion_symbol_table` structs. Ion 1.1 adds the `$ion_encoding` directive.
+
+document::
+A stream of octets conforming to either the Ion text or binary specification. Can consist of
+multiple _segments_, perhaps using varying versions of the Ion specification. A document doesn’t
+necessarily exist as a file, and isn’t necessarily finite.
+
+E-expression::
+See _encoding expression_.
+
+encoding directive::
+In Ion 1.1, a top-level `$ion_encoding` struct that controls the segment immediately following it.
+The 1.0 `$ion_symbol_table` directive is effectively a less capable alternative syntax.
+
+encoding environment::
+The context-specific data maintained by an Ion implementation while encoding or decoding data. In
+Ion 1.0 this consists of the local symbol table; in Ion 1.1 it also has the local module table
+and a collection of available modules.
+
+encoding expression::
+The invocation of a macro in encoded data, aka E-expression. Starts with a macro reference
+denoting the function to invoke. The Ion text format uses “smile syntax” `(:__macro__ …)` to denote
+E-expressions. Ion binary devotes a large number of opcodes to E-expressions, so they can be
+compact.
+
+macro::
+A transformation function that accepts some number of streams of values, and produces a stream of
+values.
+
+macro definition::
+Specifies a macro in terms of a _signature_ and a _template_.
+
+macro reference::
+Identifies a macro for invocation, renaming, or exporting. Must always be unambiguous. Lexically
+scoped, and never a “forward reference” to a macro that’s declared later in the document.
+
+module::
+The data entity that defines and exports both symbols and macros. Modules are imported by
+encoding directives then installed into the local symbol and/or macro tables.
+
+parameter::
+A named input to a macro, as defined by its signature. At expansion time a parameter produces a
+stream of values.
+
+qualified macro reference::
+A macro reference that consists of a module name and either a macro name exported by that module,
+or a numeric address within the range of the module’s exported macro table. In text, these look
+like `:__module-name__:__name-or-address__`.
+
+rest parameter::
+A macro parameter—always the final parameter—declared with the `*\...*` modifier, that accepts all
+remaining arguments to the macro. Similar to “varargs” parameters in Java and other languages.
+
+segment::
+A contiguous portion of a document that uses the same encoding environment. Segment boundaries
+are caused by directives: an IVM starts a new segment, while `$ion_symbol_table` and `$ion_encoding`
+directives end segments (with a new one starting immediately afterwards).
```

Comment:
Doesn't an IVM also end segments? (In Ion 1.0, it effectively resets the symbol table, right?)

Reply:
I didn't make myself clear. An IVM is not the end of a segment; it causes the current segment to end. So rather than make a distinction between IVMs and encoding directives, why not say something like this?

> An IVM or encoding directive _causes_ the current segment to end and _is_ the start of a new segment. 

I _think_ it's the same in practice as what you said, but it avoids special-casing them. 

(Technically, my phrasing _might_ imply that every Ion text document starts with a possibly empty Ion 1.0 segment, but I don't think that's a problem since implementations have to treat it that way anyway.)

Reply:
A segment is a non-overlapping, contiguous range of bytes, and the bytes encoding a directive lie within exactly one segment.  I'm trying to be clear about which segment a directive resides in, because that determines how that directive is encoded. 

IVMs are inherently a special case compared to other directive, because they are not encoded as data but as a special keyword (in text) or opcode (in binary).   And because `$ion_1_1` is not valid keyword in 1.0, it must necessarily be part of a new segment.

I think the statement "every Ion text document starts with a possibly empty Ion 1.0 segment" is true and reasonable, assuming that we allow for empty segments.  Consider:
```
// first line
$ion_1_1  // second line
// third line
```
The first line is within a 1.0 segment, because that's the default for Ion documents, and the second and third lines are within a 1.1 segment.

PS: I'm not fond of "causes" because it implies an active role to passive data.  I think it's better to focus on the _meaning_ (denotation) of the data and how its structured.  (I'm sure I still use active phrasing, but I try to avoid it.)